# SARS-CoV-2-HPDA-evolutionary-analysis
@Author: Arnaud N'Guessan

## Overview
This repository contains a script for analyzing SARS-CoV-2 evolution in epitopes during the first two waves of the COVID-19 pandemic. This script and the related data can be updated manually to integrate data from other waves or other sets of epitopes. 

## Dependencies
R (version 3.5.2+) packages: "ggplot2", "seqinr", "grid", "RColorBrewer", "randomcoloR", "gplots", "lmPerm", "ggpubr", "gridExtra", "RColorBrewer", "indicspecies", "tidyr", "Cairo", "parallel", "foreach", "doParallel", "infotheo", "VennDiagram", "Biostrings", "session"

## Scripts

a) Inputs (files you need to copy into the Variant_calling workspace): 

-->OUTPUT_WORKSPACE: The absolute path of the "SARS-CoV-2-HPDA-evolutionary-analysis/" repertory in your system. Make sure that it contains a sub-directory named "depth_report_NCBI_SRA_amplicon/" which contains all the samples depth coverage analysis files (a .csv file generated by "samtools depth" or a csv file with 3 columns/fields corresponding to the sample, the position of the site in the reference genome MN908947_3 and the site depth respectively). We added an example of such depth report file in "SARS-CoV-2-HPDA-evolutionary-analysis/depth_report_NCBI_SRA_amplicon/" so that you can visualize what it should look like (each sample needs to have its own depth report file). Next, make sure that the "SARS-CoV-2-HPDA-evolutionary-analysis/" repertory should also contain the script (high_confidence_epitopes_analysis.r) and the related data (Epitopes_mapped.csv, MN908947_3.fasta, Table_df_all_mutations_prevalence_in_lineages.rds, df_high_confidence_epitope_metrics.rds, df_sars_cov_2_epitopes.rds, df_variants_SRA_amplicon_first_wave.rds and df_variants_SRA_amplicon_second_wave.rds)

-->NB_CPUS: the number of cpus to use for analyzes that are performed through parrallel programming (R doParallel)

b) Various plots showing the evolutionary profile of SARS-CoV-2 epitopes during waves 1 and 2 + comparisons between lineages / variants.

For running the script from a terminal (command line), you must have R (version 3.5.2+) installed or loaded (slurm module) and you must run the command:
Rscript high_confidence_epitopes_analysis.r $OUTPUT_WORKSPACE $NB_CPUS
